### PR TITLE
Non-linear mapping, part1

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Prepares platforms to handle non-linear mapping, but waiting with the actual non-linear mapping until next P-R

### CODE REVIEW

File: core/arch/arm/plat-stm/tz_a9init.S [hunk 1]
```diff
@@ -176,15 +165,11 @@ UNWIND(	.fnstart)
...
-	movt r0, #PL310_BASE_H
 	mov  r1, #0x1
-	str  r1, [r0]
+	str  r1, [r0, #PL310_CTRL]
 
 	/* if L2 FLZW enable, enable in L1 */
-	movw r0, #PL310_AUX_CTRL
-	movt r0, #PL310_BASE_H
-	ldr  r1, [r0]
+	ldr  r1, [r0, #PL310_CTRL]
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/plat-stm/tz_a9init.S]: s/PL310_CTRL/PL310_AUX_CTRL

File: core/arch/arm/plat-stm/tz_a9init.S [hunk 2]
```diff
@@ -378,24 +351,22 @@ loop_1:
 
 
 	/*- PL310 Memory Controller (Note: should be done with NS=1) ---*/
+	/* XXX we're in secure mode with mmu disabled => NS=0 */
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/plat-stm/tz_a9init.S]: why this comment?

Comment (jenswi-linaro) [inline core/arch/arm/plat-stm/tz_a9init.S]: I can remove it, if that's what you want.

File: core/arch/arm/plat-stm/tz_a9init.S (hunk 1 above)
Comment (jenswi-linaro) [inline core/arch/arm/plat-stm/tz_a9init.S]: I'll fix.

File: core/arch/arm/plat-stm/asc.S [hunk 3]
```diff
@@ -77,16 +64,17 @@ UNWIND(	.fnend)
...
+ *    R1 is the base address of the uart
  *    Clobbers r0-r3
  */
 FUNC __asc_xmit_char , :
 UNWIND(	.fnstart)
 
-    LDR r2, =BOARD_ASC_TXBUFFER_REG
-    LDR r3, =BOARD_ASC_STATUS_REG
+    ADD r2, r0, #ASC_TXBUFFER
+    ADD r3, r0, #ASC_STATUS
```
Comment (pascal-brand-st-dev) [inline core/arch/arm/plat-stm/asc.S]: s/r0/r1  in both instruction

Comment (jenswi-linaro) [inline core/arch/arm/plat-stm/asc.S]: I'll fix

File: core/arch/arm/plat-stm/tz_a9init.S (hunk 2 above)
Comment (pascal-brand-st-dev) [inline core/arch/arm/plat-stm/tz_a9init.S]: The `XXX` is odd. But we can keep it of course

Comment (jenswi-linaro) [inline core/arch/arm/plat-stm/tz_a9init.S]: OK, I'll change to "Note" instead.

**Comment by pascal-brand-st-dev** [comment]
With the proposed fixes on plat-stm:
`Tested-by: Pascal Brand <pascal.brand@linaro.org> (STM)`
`Reviewed-by: Pascal Brand <pascal.brand@linaro.org>`

Should we ask testing on all platforms before merging it? I'm ok with either way.

**Comment by jenswi-linaro** [comment]
Updated and tags applied.

**Comment by jenswi-linaro** [comment]
@jbech-linaro, what do you say about testing the other platforms?

**Comment by jforissier** [comment]
`Tested-by: Jerome Forissier <jerome.forissier@linaro.org> (HiKey)`

For the record: tested 32 bits, 64 bits and 32 bits + `CFG_WITH_PAGER=y`
```
$ xtest
[...]
26853 subtests of which 0 failed
721 test cases of which 0 failed
0 test case was skipped
```

**Note**: I got one crash when CFG_WITH_PAGER=y, which does not seem to be related to this PR because I could reproduce it on `master` (without this PR). It does not occur always. I am investigating further and will open a bug report if it's not a mistake on my part.
```
o XTEST_TEE_1013.1.5 Busy loop repeat 10000
ERROR:   TEE-CORE:
ERROR:   TEE-CORE: core data-abort at address 0x6 (alignment fault)
ERROR:   TEE-CORE:  fsr 0x00000801  ttbr0 0x3f0104ea  ttbr1 0x3f00c06a  cidr 0x1
ERROR:   TEE-CORE:  cpu #4          cpsr 0x800000b3
ERROR:   TEE-CORE:  r0 0x3f03edec      r4 0x00000001    r8 0x00115b90   r12 0x00100fb8
ERROR:   TEE-CORE:  r1 0x3f03edf0      r5 0x00100fa0    r9 0x00000000    sp 0x3f03edec
ERROR:   TEE-CORE:  r2 0x0010c0a0      r6 0x00000000   r10 0x00000002    lr 0x3f000f20
ERROR:   TEE-CORE:  r3 0x0010c0a8      r7 0x00000000   r11 0x3f03eebc    pc 0x3f01d70a
ERROR:   TEE-CORE: [abort] alignement fault!  (trap CPU)
ERROR:   TEE-CORE: PANIC: get_fault_type core/arch/arm/kernel/abort.c:512
```

**Comment by jenswi-linaro** [comment]
Interesting that you got this problem, I've also seen it from time to time. But never without some additional changes.

**Comment by jforissier** [comment]
I have entered https://github.com/OP-TEE/optee_os/issues/794. Quite easy to reproduce with CFG_NUM_THREADS=8.

**Comment by jbech-linaro** [comment]
@jenswi-linaro , it's OK to merge this as it is now. We have a the next release in about a month and in the beginning of June we should start/prepare to re-test everything again for the release and then we will find out where it's not working (if that is the case).

**Comment by jbech-linaro** [comment]
:+1:  `Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

**Comment by jenswi-linaro** [comment]
Rebased, tags applied.

**Pull request merged**